src/dbnode/storage/profiler/profiler.go

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package profiler contains the code used for profiling.
+package profiler
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime/pprof"
+	"sync/atomic"
+)
+
+const (
+	// ProfileFileExtension is the extension of profile files.
+	ProfileFileExtension = ".pb.gz"
+)
+
+// FileProfileContext is file profiler context data.
+type FileProfileContext struct {
+	path        string
+	profileName *profileName
+}
+
+// StopProfile stops started profile.
+func (f FileProfileContext) StopProfile() error {
+	stopCPUProfile()
+	return writeHeapProfile(f.path, f.profileName)
+}
+
+// FileProfiler is profiler which writes its profiles to given path directory.
+type FileProfiler struct {
+	path         string
+	profileNames map[string]*profileName
+}
+
+// NewFileProfiler creates a new file provider.
+func NewFileProfiler(path string) *FileProfiler {
+	return &FileProfiler{
+		path:         path,
+		profileNames: make(map[string]*profileName),
+	}
+}
+
+// StartProfile starts a new named profile.
+func (f FileProfiler) StartProfile(name string) (ProfileContext, error) {
```

Comment:
In my opinion, it would be clearer to simply have `StartCPUProfile()`, `StopCPUProfile()` and `WriteHeapDump()` methods and use them directly. Also, returning a `ProfileContext` might make one think that it's possible to run multiple profilings simultaneously.

Reply:
Makes sense, will change this.